runtime/lua/vim/termcap.lua

```diff
@@ -0,0 +1,86 @@
+local M = {}
+
+--- Hex encode a string.
+---
+--- @param str string String to encode
+--- @return string Hex encoded string
+local function hexencode(str)
+  local bytes = { str:byte(1, #str) }
+  local enc = {} ---@type string[]
+  for i = 1, #bytes do
+    enc[i] = string.format('%02X', bytes[i])
+  end
+  return table.concat(enc)
+end
+
+--- Hex decode a string.
+---
+--- @param enc string String to decode
+--- @return string Decoded string
+local function hexdecode(enc)
+  assert(#enc % 2 == 0, 'string must have an even number of hex characters')
+  local str = {} ---@type string[]
+  for i = 1, #enc, 2 do
+    local n = assert(tonumber(string.format('0x%s', enc:sub(i, i + 1))))
+    str[#str + 1] = string.char(n)
+  end
+  return table.concat(str)
+end
+
+--- Query the host terminal emulator for terminfo capabilities.
+---
+--- This function sends the XTGETTCAP DCS sequence to the host terminal emulator asking the terminal
+--- to send us its terminal capabilities. These are strings that are normally taken from a terminfo
+--- file, however an up to date terminfo database is not always available (particularly on remote
+--- machines), and many terminals continue to misidentify themselves or do not provide their own
+--- terminfo file, making the terminfo database unreliable.
+---
+--- Querying the terminal guarantees that we get a truthful answer, but only if the host terminal
+--- emulator supports the XTGETTCAP sequence.
```

Comment:
Out of scope here, but if the terminal supports XTGETTCAP we should ignore terminfo entirely (by default at least).

Reply:
We get _a lot_ of information out of terminfo. Getting all of that via XTGETTCAP would result in a lot of overhead.

Terminfo has a bad reputation but most of the newer wave of terminal emulators all ship with their own terminfo entries which are actually correct (kitty, Wezterm, foot, Ghostty, and Alacritty all do). The worst offenders are those that masquerade as `xterm-256color`. I would be more open to ignoring terminfo when `$TERM` is `xterm-256color`

Reply:
I guess that means XTGETTCAP doesn't have a "batch" request which returns everything?

Reply:
No, we have to request each capability that we want in the request. The response differs between terminal emulators: Xterm returns all of the requested capabilities in a single response, while foot, Ghostty, kitty, and Wezterm (not sure about others) send separate DCS responses for each capability.